src/backend/lib/async-service.js

```diff
@@ -0,0 +1,199 @@
+  #stopped = {
+    promise: Promise.resolve(),
+    resolve: () => {},
+    reject: () => {},
+  };
+
+  /** @param {AsyncServiceStateDefault} newState */
+  setState(newState) {
+    this.#state = newState;
+  }
+
+  /** @returns {AsyncServiceStateDefault} */
+  getState() {
+    return this.#state;
+  }
+
+  /**
+   * To be overridden by implementing class
+   *
+   * @param {any[]} [args]
+   */
+  async _start(...args) {}
+  /**
+   * To be overridden by implementing class
+   */
+  async _stop() {}
+
+  /**
+   * Will resolve when the service is in started state. E.g. to ensure an async
+   * method only runs when the service is in "started" state, use:
+   *
+   * ```js
+   * await this.started()
+   * ```
+   *
+   * Will reject if the service is in "error" state.
+   *
+   * Note: If the service is in "stopping" or "stopped" state this will queue
+   * until the next time the service starts. If this is not desirable behaviour,
+   * check this.getState().value first
+   *
+   * @readonly
+   * @returns {Promise<void>}
+   */
+  async started() {
+    return this.#started.promise;
+  }
+
+  /**
+   * Will resolve when the service is in stopped state. Less useful than
+   * `started()` E.g. to ensure an async method only runs when the service is in
+   * "stopped" state, use:
+   *
+   * ```js
+   * await this.stopped()
+   * ```
+   *
+   * Will reject if the service is in "error" state.
+   *
+   * Note: If the service is in "starting" or "started" state this will queue
+   * until the next time the service stops. If this is not desirable behaviour,
+   * check this.getState().value first
+   *
+   * @readonly
+   * @returns {Promise<void>}
+   */
+  async stopped() {
+    return this.#started.promise;
+  }
+
+  /**
+   * Start service. If the service is starting or started, will resolve when the
+   * service is started, and will not call _start() for than once. If the
+   * service is in the process of stopping, will wait until it stops before
+   * starting and will not call _stop() more than once
+   *
+   * @readonly
+   * @param {any[]} [args]
+   * @returns {Promise<void>} Resolves when service is started
+   */
+  async start(...args) {
+    switch (this.getState().value) {
+      case "starting":
+      case "started":
+        // Will still resolve once has started
+        return this.#started.promise;
+      case "error":
+        this.#started.reject(this.getState().error);
+        return this.#started.promise;
+      case "stopping":
+        // Wait until stopped before continuing
+        await this.#stopped.promise;
+        break;
+      case "stopped":
+        break;
+      default:
+        return this.#started.reject(new Error("Should not happen"));
+    }
+    // We've begun starting, so stopping promise needs to be unresolved until
+    // we stop again
+    this.#stopped = pDefer();
+    try {
+      this.setState({ value: "starting" });
+      await this._start.apply(this, args);
```

Comment:
We can only get here if `state.value === "stopped"` (because of switch above).
If `state.value === "stopped"` we guarantee that `#started.promise` is unfulfilled (e.g. pending) because the action `stopping` --> `stopped` (the only way to get to the `stopped` state) resets that promise.
The reset of `#started` needs to happen there, rather than here, because otherwise multiple calls to `start()` in the same tick would return different promises, and we would need to track and resolve all of them.
At least that all makes sense in my head! I tried to make sure the tests check all of this.

Reply:
I haven't fully thought this through, but: what if `setState` was the only part that touched the started/stopped promises? That would centralize the logic & maybe make it easier to reason about.

Reply:
Added to this PR's task list